inst/include/cpp11/matrix.hpp

```diff
@@ -86,24 +191,32 @@ class matrix {
 
   r_vector<r_string> names() const { return SEXP(vector_.names()); }
 
-  row operator[](const int pos) { return {*this, pos}; }
-
-  T operator()(int row, int col) { return vector_[row + (col * nrow_)]; }
+  T at(int row, int col) const { return vector_[row + (col * nrow())]; }
+  T operator()(int row, int col) const { return at(row, col); }
+  slice operator[](int index) { return {*this, index}; }
 
-  row begin() { return {*this, 0}; }
-  row end() { return {*this, nrow_}; }
+  slice_iterator begin() { return {*this, 0}; }
+  slice_iterator end() { return {*this, nslices()}; }
 };
 
-using doubles_matrix = matrix<r_vector<double>, double>;
-using integers_matrix = matrix<r_vector<int>, int>;
-using logicals_matrix = matrix<r_vector<r_bool>, r_bool>;
-using strings_matrix = matrix<r_vector<r_string>, r_string>;
+template <typename S = by_column>
```

Comment:
I think the default access pattern would need to be by row, as this is what cpp11 currently supplies.

Reply:
I can change that. I just have thought that cpp11 is in active development and have not reached 1.0.0 yet, so changes like this are still possible. The reason is - column slices have continuous memory layout, so it would be nice to select more efficient alternative by default. And the fact that there were some issues with matrix iterators suggests that so far there are not so many matrix users whose code might be broken.

Reply:
That is true, let me ask around and see what people might prefer.

Reply:
Alright since R generally does column major ordering by default I think having cpp11 match that makes the most sense. So I think leaving it as you have it now is OK

Reply:
Great! Thanks for double-checking.
I've applied some cosmetics and squashed the commits to clean-up the log.